Mathlib/Probability/Kernel/Disintegration/Integral.lean

```diff
@@ -0,0 +1,280 @@
+  [MeasurableSpace Ω] [StandardBorelSpace Ω] [Nonempty Ω]
+
+section Lintegral
+
+variable [CountableOrCountablyGenerated α β] {ρ : Measure (β × Ω)} [IsFiniteMeasure ρ]
+  {f : β × Ω → ℝ≥0∞}
+
+lemma lintegral_condKernel_mem {s : Set (β × Ω)} (hs : MeasurableSet s) :
+    ∫⁻ x, ρ.condKernel x {y | (x, y) ∈ s} ∂ρ.fst = ρ s := by
+  conv_rhs => rw [← compProd_fst_condKernel ρ]
+  simp_rw [compProd_apply hs]
+  rfl
+#align probability_theory.lintegral_cond_kernel_mem MeasureTheory.Measure.lintegral_condKernel_mem
+
+lemma set_lintegral_condKernel_eq_measure_prod {s : Set β} (hs : MeasurableSet s) {t : Set Ω}
+    (ht : MeasurableSet t) :
+    ∫⁻ b in s, ρ.condKernel b t ∂ρ.fst = ρ (s ×ˢ t) := by
+  have : ρ (s ×ˢ t) = (ρ.fst ⊗ₘ ρ.condKernel) (s ×ˢ t) := by
+    congr; exact (compProd_fst_condKernel ρ).symm
+  rw [this, compProd_apply (hs.prod ht)]
+  classical
+  have : ∀ b, ρ.condKernel b (Prod.mk b ⁻¹' s ×ˢ t)
+      = s.indicator (fun b ↦ ρ.condKernel b t) b := by
+    intro b
+    by_cases hb : b ∈ s <;> simp [hb]
+  simp_rw [this]
+  rw [lintegral_indicator _ hs]
+#align probability_theory.set_lintegral_cond_kernel_eq_measure_prod MeasureTheory.Measure.set_lintegral_condKernel_eq_measure_prod
+
+lemma lintegral_condKernel (hf : Measurable f) :
+    ∫⁻ b, ∫⁻ ω, f (b, ω) ∂(ρ.condKernel b) ∂ρ.fst = ∫⁻ x, f x ∂ρ := by
+  conv_rhs => rw [← compProd_fst_condKernel ρ]
+  rw [lintegral_compProd hf]
+#align probability_theory.lintegral_cond_kernel MeasureTheory.Measure.lintegral_condKernel
+
+lemma set_lintegral_condKernel (hf : Measurable f) {s : Set β}
+    (hs : MeasurableSet s) {t : Set Ω} (ht : MeasurableSet t) :
+    ∫⁻ b in s, ∫⁻ ω in t, f (b, ω) ∂(ρ.condKernel b) ∂ρ.fst
+      = ∫⁻ x in s ×ˢ t, f x ∂ρ := by
+  conv_rhs => rw [← compProd_fst_condKernel ρ]
+  rw [set_lintegral_compProd  hf hs ht]
+#align probability_theory.set_lintegral_cond_kernel MeasureTheory.Measure.set_lintegral_condKernel
+
+lemma set_lintegral_condKernel_univ_right (hf : Measurable f) {s : Set β}
+    (hs : MeasurableSet s) :
+    ∫⁻ b in s, ∫⁻ ω, f (b, ω) ∂(ρ.condKernel b) ∂ρ.fst
+      = ∫⁻ x in s ×ˢ Set.univ, f x ∂ρ := by
+  rw [← set_lintegral_condKernel hf hs MeasurableSet.univ]; simp_rw [Measure.restrict_univ]
+#align probability_theory.set_lintegral_cond_kernel_univ_right MeasureTheory.Measure.set_lintegral_condKernel_univ_right
+
+lemma set_lintegral_condKernel_univ_left (hf : Measurable f) {t : Set Ω}
+    (ht : MeasurableSet t) :
+    ∫⁻ b, ∫⁻ ω in t, f (b, ω) ∂(ρ.condKernel b) ∂ρ.fst
+      = ∫⁻ x in Set.univ ×ˢ t, f x ∂ρ := by
+  rw [← set_lintegral_condKernel hf MeasurableSet.univ ht]; simp_rw [Measure.restrict_univ]
+#align probability_theory.set_lintegral_cond_kernel_univ_left MeasureTheory.Measure.set_lintegral_condKernel_univ_left
+
+end Lintegral
+
+section Integral
+
+variable {ρ : Measure (β × Ω)} [IsFiniteMeasure ρ]
+  {E : Type*} {f : β × Ω → E} [NormedAddCommGroup E] [NormedSpace ℝ E] [CompleteSpace E]
```

Comment:
It seems to me that the `[CompleteSpace E]` assumption shouldn't be needed: if the space is not complete, then all integrals vanish by our choice of junk values, so all the theorems below remain true. I've taken the liberty to push a commit removing some of these assumptions here and there.

Reply:
Great, thanks!